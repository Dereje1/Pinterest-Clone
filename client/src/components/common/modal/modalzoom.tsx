// displays pin zoom modal
import React, { Component, createRef } from 'react';
/* MUI */
import Card from '@mui/material/Card';
import { CardContent } from '@mui/material';
/* local components and utility */
import ModalHeader from './Header/ModalHeader';
import Comments from './Back/Comments';
import SwipableImage from './Front/SwipableImage';
import { delay } from '../../../utils/utils';
import {
  PinType, userType, zoomedImageInfoType, imageMetadataType,
} from '../../../interfaces';
import './modal.scss';

interface PinZoomProps {
  zoomInfo: zoomedImageInfoType,
  user: userType
  reset: () => void
  pinImage: (pin: PinType) => void
  deletePin: ((pin: PinType) => void) | null
  handleNewComment: (newComment: string) => void
  updateTags: (query: string) => void
  displayLogin: () => void
  onSlidePin: (newIndex: number) => void
  resetParentDivStyle: (metadata: imageMetadataType | null) => zoomedImageInfoType['parentDivStyle'] | null
}

interface PinZoomState {
  commentsShowing: { width: number, height: number } | null
  cancelBlur: boolean
  zoomClass: string
  imgMetaData: imageMetadataType | null
}

// The PinZoom component displays a zoomed-in view of a pin in a modal.
// It also provides functionality to view and add comments to the pin.
export class PinZoom extends Component<PinZoomProps, PinZoomState> {

  // Reference to the zoomed image element
  private zoomedImage = createRef<HTMLDivElement>();

  constructor(props: PinZoomProps) {
    super(props);
    this.state = {
      commentsShowing: null,
      cancelBlur: false,
      zoomClass: 'zoom cshow',
      imgMetaData: null,
    };
  }

  // Focus on the zoomed image when the component mounts
  componentDidMount() {
    if (this.zoomedImage.current) {
      this.zoomedImage.current.focus();
    }
  }

  // Close the modal and reset the state
  close = (_: React.SyntheticEvent, forceClose = false) => {
    const { cancelBlur } = this.state;
    if (cancelBlur && !forceClose) return;
    const { reset } = this.props;
    this.setState({
      zoomClass: 'zoom chide',
    }, async () => {
      await delay(500);
      reset();
    });
  };

  // Calculate and set the state for showing comments
  handleCommentsNotShowing = (zoomedImage: HTMLDivElement, updatedDivStyle: zoomedImageInfoType['parentDivStyle']) => {
    const [header] = zoomedImage.children;
    this.setState({
      commentsShowing: {
        width: updatedDivStyle.parentWidth,
        height: window.innerHeight - header.clientHeight - 25,
      },
      cancelBlur: true,
    });
  };

  // Handle the case when comments are currently showing
  handleCommentsShowing = (zoomedImage: HTMLDivElement) => {
    zoomedImage.focus();
    this.setState({ commentsShowing: null, cancelBlur: false });
  };

  // Toggle the visibility of the comments section
  toggleComments = () => {
    const { commentsShowing, imgMetaData } = this.state;
    const { resetParentDivStyle } = this.props;
    const updatedDivStyle = resetParentDivStyle(imgMetaData);
    if (this.zoomedImage.current && updatedDivStyle) {
      if (!commentsShowing) {
        this.handleCommentsNotShowing(this.zoomedImage.current, updatedDivStyle);
      } else {
        this.handleCommentsShowing(this.zoomedImage.current);
      }
    }
  };

  // Render the PinZoom component
  render() {
    const {
      zoomInfo: { pin: pinInformation, parentDivStyle },
      pinImage,
      deletePin,
      user: { authenticated },
      handleNewComment,
      updateTags,
      displayLogin,
      zoomInfo,
      onSlidePin,
    } = this.props;
    const {
      commentsShowing, zoomClass,
    } = this.state;

    return (
      <>
        <div className="modal-overlay" />
        <Card
          sx={{
            width: parentDivStyle.parentWidth,
            top: parentDivStyle.top,
            background: commentsShowing ? '' : 'rgba(0, 0, 0, 0.7)',
          }}
          className={zoomClass}
          onBlur={this.close}
          ref={this.zoomedImage}
          tabIndex={0}
        >
          <ModalHeader
            authenticated={authenticated}
            commentsShowing={Boolean(commentsShowing)}
            pinInformation={pinInformation}
            closePin={this.close}
            deletePin={deletePin}
            displayLogin={displayLogin}
            pinImage={pinImage}
            toggleComments={this.toggleComments}
          />
          <CardContent sx={{
            background: parentDivStyle.isNoFit && !commentsShowing ? 'black' : '',
            padding: 0,
            '&:last-child': { pb: 0 },
          }}
          >
            {!commentsShowing ? (
              <SwipableImage
                zoomInfo={zoomInfo}
                onSlidePin={onSlidePin}
                onSetImageMetaData={(m) => this.setState({ imgMetaData: m })}
              />
            )
              : (
                <Comments
                  stylingProps={commentsShowing}
                  pinInformation={pinInformation}
                  handleNewComment={handleNewComment}
                  authenticated={authenticated}
                  toggleComments={this.toggleComments}
                  closePin={(e: React.SyntheticEvent) => this.close(e, true)}
                  updateTags={updateTags}
                  displayLogin={displayLogin}
                />
              )}
          </CardContent>
        </Card>
      </>
    );
  }

}

export default PinZoom;
