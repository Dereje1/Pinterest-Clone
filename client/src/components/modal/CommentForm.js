import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@mui/material/Button';
import TextField from '@mui/material/TextField';

function CommentForm({ handleClose, handleSubmit }) {
  const [comment, setComment] = useState('');

  return (
    <div style={{
      margin: 5, display: 'flex', flexDirection: 'column', alignItems: 'flex-end',
    }}
    >
      <TextField
        autoFocus
        id="comment_form"
        label="Add a comment"
        type="text"
        multiline
        rows={3}
        fullWidth
        value={comment}
        onChange={(e) => setComment(e.target.value)}
      />
      <div style={{ marginTop: 1 }}>
        <Button onClick={handleClose}>Cancel</Button>
        <Button onClick={() => handleSubmit(comment)}>Submit</Button>
      </div>
    </div>
  );
}

export default CommentForm;

CommentForm.propTypes = {
  handleClose: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
};
