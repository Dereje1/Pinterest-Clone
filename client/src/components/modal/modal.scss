@import "./../../common.scss";

.zoom{
  height: auto;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  margin-left: auto;
  margin-right: auto;
  background: white;
  overflow: auto;
  box-shadow: 0 5px 15px rgba(0,0,0,.5);
  border: 1px solid rgba(0,0,0,.2);
  border-radius: 6px;
  z-index: 1010;
  outline: none;
  &.cshow{
    display: block;
    @include animation('zoom-pin');
    animation-duration: 1s;
    animation-iteration-count: 1;
    animation-fill-mode: forwards;
  }
  &.chide{
    display: block;
    @include animation('restore-pin');
    animation-duration: 0.5s;
    animation-iteration-count: 1;
    animation-fill-mode: forwards;
  }
}


@include keyframes(zoom-pin) {
  0%{
    transform: scale(0.6);
  }
  100%{
    transform: scale(1);
    visibility: visible;
  }
}

@include keyframes(restore-pin) {
  0%{
    transform: scale(1);
  }
  100%{
    transform: scale(0);
    visibility:hidden;
  }
}

