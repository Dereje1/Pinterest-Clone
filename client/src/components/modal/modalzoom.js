// displays pin zoom modal
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Card from '@mui/material/Card';
import Tooltip from '@mui/material/Tooltip';
import CardHeader from '@mui/material/CardHeader';
import CardMedia from '@mui/material/CardMedia';
import Badge from '@mui/material/Badge';
import PinDropIcon from '@mui/icons-material/PinDrop';
import { styled } from '@mui/styles';
import ModalActions from './ModalActions';
import {
  delay, getNewImageWidth, getPinners, getFormattedDescription,
} from '../../utils/utils';
import './modal.scss';

const StyledBadge = styled(Badge)(() => ({
  '& .MuiBadge-badge': {
    right: 16,
    top: 0,
    border: '2px solid grey',
    padding: '0 4px',
  },
}));

export class PinZoom extends Component {

  constructor(props) {
    super(props);
    // initialize modal show state to false
    this.state = {
      show: false,
      parentDivStyle: { top: 0, width: '90%' },
    };
    this.zoomedImage = React.createRef();
  }

  componentDidMount() {
    this.zoomedImage.current.focus();
    const { zoomInfo } = this.props;
    const { parentDivStyle } = this.state;
    const [, browserTop] = zoomInfo;
    // set top and scroll to current position and disable scroll
    window.scrollTo(0, browserTop);
    document.body.style.overflow = 'hidden';
    this.setState({
      show: true,
      parentDivStyle: {
        ...parentDivStyle,
        top: browserTop,
      },
    }, async () => { await delay(1000); });
  }

  componentWillUnmount() {
    document.body.style.overflowY = 'scroll';
  }

  close = () => {
    // sends a reset callback after closing modalstate
    const { reset } = this.props;
    this.setState({
      show: false,
    }, async () => {
      // delay to display closing keyframe
      await delay(500);
      reset();
    });
  };

  handleImage = ({ target: { naturalWidth, naturalHeight } }) => {
    const { parentDivStyle } = this.state;
    const newDivStyle = getNewImageWidth({ naturalWidth, naturalHeight });
    // const pcopy = JSON.parse(JSON.stringify(parentDivStyle));
    this.setState({
      parentDivStyle: {
        ...parentDivStyle,
        ...newDivStyle,
      },
    });
  };


  render() {
    const {
      zoomInfo, pinImage, deletePin,
    } = this.props;
    const { show, parentDivStyle } = this.state;
    if (!zoomInfo.length) return null;
    const [pinInformation] = zoomInfo;
    const totalPins = (pinInformation.savedBy) ? pinInformation.savedBy.length : 0;
    const pinnedBy = totalPins ? getPinners(pinInformation.savedBy) : '';
    const formattedDescription = getFormattedDescription(pinInformation.imgDescription);
    const [, day, mth, year] = new Date(pinInformation.createdAt).toUTCString().split(' ');
    return (
      <>
        <div className="modal-overlay" />
        <Card
          sx={parentDivStyle}
          className={show ? 'zoom cshow' : 'zoom chide'}
          onBlur={this.close}
          ref={this.zoomedImage}
          tabIndex={0}
        >
          <CardHeader
            action={(
              <>
                <ModalActions
                  element={pinInformation}
                  pinImage={pinImage}
                  deletePin={deletePin}
                  reset={this.close}
                />
                <StyledBadge badgeContent={totalPins} color="secondary" showZero>
                  <Tooltip title={pinnedBy}>
                    <PinDropIcon style={{ fontSize: parentDivStyle.pinnersSize }} />
                  </Tooltip>
                </StyledBadge>
              </>
            )}
            title={formattedDescription}
            subheader={(
              <>
                <span>
                  {pinInformation.owner}
                </span>
                <br />
                <span style={{ fontSize: parentDivStyle.dateSize, marginLeft: 0 }}>
                  {`${day} ${mth} ${year}`}
                </span>
              </>
            )}
            titleTypographyProps={{ fontSize: parentDivStyle.titleSize, fontWeight: 'bold' }}
            subheaderTypographyProps={{ fontSize: parentDivStyle.subTitleSize, fontWeight: 'bold' }}
          />
          <CardMedia
            component="img"
            image={pinInformation.imgLink}
            onLoad={this.handleImage}
            id="pin-zoom"
          />
        </Card>
      </>
    );
  }

}

export default PinZoom;


PinZoom.defaultProps = {
  pinImage: null,
  deletePin: null,
};

PinZoom.propTypes = {
  // [picobject, overlay button type, last scroll distance]
  zoomInfo: PropTypes.arrayOf(PropTypes.any).isRequired,
  // callback to caller to turn modal off
  reset: PropTypes.func.isRequired,
  // what type of button to place on pic/thumbnail executed by caller
  pinImage: PropTypes.func,
  deletePin: PropTypes.func,
};
