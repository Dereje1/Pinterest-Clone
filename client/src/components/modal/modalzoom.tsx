// displays pin zoom modal
import React, { Component, createRef } from 'react';
import { Link as RouterLink } from 'react-router-dom';
/* MUI */
import Card from '@mui/material/Card';
import CardHeader from '@mui/material/CardHeader';
import { CardContent } from '@mui/material';
import CardMedia from '@mui/material/CardMedia';
import Badge from '@mui/material/Badge';
import IconButton from '@mui/material/IconButton';
import CommentOutlinedIcon from '@mui/icons-material/CommentOutlined';
import CommentIcon from '@mui/icons-material/Comment';
import Link from '@mui/material/Link';
import { styled } from '@mui/styles';
/* local components and utility */
import ModalActions from './ModalActions';
import Comments from './Comments';
import {
  delay, getFormattedDescription, formatDate,
} from '../../utils/utils';
import { PinType, userType, zoomedImageInfoType } from '../../interfaces';
import './modal.scss';

export const StyledBadge = styled(Badge)(({ name }:{ name: string}) => ({
  '& .MuiBadge-badge': {
    right: name === 'pin' ? 32 : 43,
    top: name === 'pin' ? 17 : 13,
    border: '2px solid grey',
    padding: '0 4px',
  },
}));

interface PinZoomProps {
    zoomInfo: zoomedImageInfoType,
    user: userType
    reset: () => void
    pinImage: (pin: PinType) => void
    deletePin: ((pin: PinType) => void) | null
    handleNewComment: (newComment: string) => void
    updateTags: (query: string) => void
}

interface PinZoomState {
  commentsShowing: {width: number, height: number} | null
  cancelBlur: boolean
  zoomClass: string
}

export class PinZoom extends Component<PinZoomProps, PinZoomState> {

  private zoomedImage = createRef<HTMLDivElement>();

  constructor(props: PinZoomProps) {
    super(props);
    this.state = {
      commentsShowing: null,
      cancelBlur: false,
      zoomClass: 'zoom cshow',
    };
  }

  componentDidMount() {
    if (this.zoomedImage.current) {
      this.zoomedImage.current.focus();
    }
    window.addEventListener('scroll', this.disableScroll);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.disableScroll);
  }

  disableScroll = () => {
    const { zoomInfo: { parentDivStyle } } = this.props;
    window.scrollTo(0, parentDivStyle.top);
  };

  close = (_: React.SyntheticEvent, forceClose = false) => {
    // sends a reset callback after closing modalstate
    const { cancelBlur } = this.state;
    if (cancelBlur && !forceClose) return;
    const { reset } = this.props;
    this.setState({
      zoomClass: 'zoom chide',
    }, async () => {
      // delay to display closing keyframe
      await delay(500);
      reset();
    });
  };

  toggleComments = () => {
    const { commentsShowing } = this.state;
    const { zoomInfo: { parentDivStyle } } = this.props;
    if (this.zoomedImage.current && !commentsShowing) {
      const { current: { clientHeight: cardHeight, children } } = this.zoomedImage;
      const [, image] = children;
      const { clientHeight: imageHeight } = image;
      this.setState({
        commentsShowing: {
          width: parentDivStyle.parentWidth,
          height: imageHeight + (window.innerHeight - cardHeight - 25),
        },
        cancelBlur: true,
      });
      return;
    }
    // need to reinstate focus for blur to work again
    if (this.zoomedImage.current) {
      this.zoomedImage.current.focus();
    }
    this.setState({ commentsShowing: null, cancelBlur: false });
  };

  render() {
    const {
      zoomInfo: { pin: pinInformation, parentDivStyle },
      pinImage,
      deletePin,
      user: { authenticated },
      handleNewComment,
      updateTags,
    } = this.props;
    const {
      commentsShowing, zoomClass,
    } = this.state;
    const formattedDescription = getFormattedDescription(pinInformation.imgDescription);
    return (
      <>
        <div className="modal-overlay" />
        <Card
          sx={{
            width: parentDivStyle.parentWidth,
            top: parentDivStyle.top,
          }}
          className={zoomClass}
          onBlur={this.close}
          ref={this.zoomedImage}
          tabIndex={0}
        >
          <CardHeader
            action={(
              <>
                <StyledBadge badgeContent={pinInformation.comments.length} color="primary" showZero name="comments">
                  <IconButton
                    onClick={this.toggleComments}
                    onMouseDown={(e) => e.preventDefault()}
                  >
                    {commentsShowing
                      ? <CommentIcon style={{ fontSize: '1.7em' }} />
                      : <CommentOutlinedIcon style={{ fontSize: '1.7em' }} />}
                  </IconButton>
                </StyledBadge>
                <StyledBadge badgeContent={pinInformation.savedBy.length} color="secondary" showZero name="pin">
                  <ModalActions
                    element={pinInformation}
                    pinImage={pinImage}
                    deletePin={deletePin}
                    reset={this.close}
                  />
                </StyledBadge>
              </>
            )}
            title={formattedDescription}
            subheader={(
              <>
                <Link
                  component={RouterLink}
                  underline="none"
                  to={
                    `/profile/${pinInformation.owner.userId}`
                  }
                  onMouseDown={(e) => e.preventDefault()}
                  onClick={(e) => this.close(e, true)}
                >
                  {pinInformation.owner.name}
                </Link>
                <br />
                <span style={{ marginLeft: 0 }}>
                  {formatDate(pinInformation.createdAt)}
                </span>
              </>
            )}
            titleTypographyProps={{ fontWeight: 'bold' }}
            subheaderTypographyProps={{ fontWeight: 'bold' }}
          />
          <CardContent sx={{
            background: parentDivStyle.isNoFit && !commentsShowing ? 'black' : '',
            padding: 0,
            '&:last-child': { pb: 0 },
          }}
          >
            { !commentsShowing ? (
              <CardMedia
                component="img"
                image={pinInformation.imgLink}
                id="pin-zoom"
                sx={{
                  width: parentDivStyle.imgWidth,
                  marginLeft: 'auto',
                  marginRight: 'auto',
                }}
              />
            )
              : (
                <Comments
                  stylingProps={commentsShowing}
                  pinInformation={pinInformation}
                  handleNewComment={handleNewComment}
                  authenticated={authenticated}
                  toggleComments={this.toggleComments}
                  closePin={(e: React.SyntheticEvent) => this.close(e, true)}
                  updateTags={updateTags}
                />
              )}
          </CardContent>

        </Card>
      </>
    );
  }

}

export default PinZoom;
