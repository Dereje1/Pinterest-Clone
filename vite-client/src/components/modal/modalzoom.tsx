// displays pin zoom modal
import React, { Component, createRef } from 'react';
/* MUI */
import Card from '@mui/material/Card';
import { CardContent } from '@mui/material';
/* local components and utility */
import ModalHeader from './Header/ModalHeader';
import Comments from './Back/Comments';
import SwipableImage from './Front/SwipableImage';
import { delay } from '../../utils/utils';
import {
  PinType, userType, zoomedImageInfoType, imageMetadataType,
} from '../../interfaces';
import './modal.scss';

interface PinZoomProps {
  zoomInfo: zoomedImageInfoType,
  user: userType
  reset: () => void
  pinImage: (pin: PinType) => void
  deletePin: ((pin: PinType) => void) | null
  handleNewComment: (newComment: string) => void
  updateTags: (query: string) => void
  displayLogin: () => void
  onSlidePin: (newIndex: number) => void
  resetParentDivStyle: (metadata: imageMetadataType | null) => zoomedImageInfoType['parentDivStyle'] | null
}

interface PinZoomState {
  commentsShowing: { width: number, height: number } | null
  cancelBlur: boolean
  zoomClass: string
  imgMetaData: imageMetadataType | null
}

export class PinZoom extends Component<PinZoomProps, PinZoomState> {

  private zoomedImage = createRef<HTMLDivElement>();

  constructor(props: PinZoomProps) {
    super(props);
    this.state = {
      commentsShowing: null,
      cancelBlur: false,
      zoomClass: 'zoom cshow',
      imgMetaData: null,
    };
  }

  componentDidMount() {
    if (this.zoomedImage.current) {
      this.zoomedImage.current.focus();
    }
    window.addEventListener('scroll', this.disableScroll);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.disableScroll);
  }

  disableScroll = () => {
    const { zoomInfo: { parentDivStyle } } = this.props;
    window.scrollTo(0, parentDivStyle.top);
  };

  close = (_: React.SyntheticEvent, forceClose = false) => {
    // sends a reset callback after closing modalstate
    const { cancelBlur } = this.state;
    if (cancelBlur && !forceClose) return;
    const { reset } = this.props;
    this.setState({
      zoomClass: 'zoom chide',
    }, async () => {
      // delay to display closing keyframe
      await delay(500);
      reset();
    });
  };

  toggleComments = () => {
    const { commentsShowing, imgMetaData } = this.state;
    const { resetParentDivStyle } = this.props;
    const updatedDivStyle = resetParentDivStyle(imgMetaData);
    if (this.zoomedImage.current && updatedDivStyle) {
      if (!commentsShowing) {
        const { current: { children } } = this.zoomedImage;
        const [header] = children;
        this.setState({
          commentsShowing: {
            width: updatedDivStyle.parentWidth,
            height: window.innerHeight - header.clientHeight - 25,
          },
          cancelBlur: true,
        });
        return;
      }
      // need to reinstate focus for blur to work again
      this.zoomedImage.current.focus();
      this.setState({ commentsShowing: null, cancelBlur: false });
    }
  };

  render() {
    const {
      zoomInfo: { pin: pinInformation, parentDivStyle },
      pinImage,
      deletePin,
      user: { authenticated },
      handleNewComment,
      updateTags,
      displayLogin,
      zoomInfo,
      onSlidePin,
    } = this.props;
    const {
      commentsShowing, zoomClass,
    } = this.state;

    return (
      <>
        <div className="modal-overlay" />
        <Card
          sx={{
            width: parentDivStyle.parentWidth,
            top: parentDivStyle.top,
            background: commentsShowing ? '' : 'rgba(0, 0, 0, 0.7)',
          }}
          className={zoomClass}
          onBlur={this.close}
          ref={this.zoomedImage}
          tabIndex={0}
        >
          <ModalHeader
            authenticated={authenticated}
            commentsShowing={Boolean(commentsShowing)}
            pinInformation={pinInformation}
            closePin={this.close}
            deletePin={deletePin}
            displayLogin={displayLogin}
            pinImage={pinImage}
            toggleComments={this.toggleComments}
          />
          <CardContent sx={{
            background: parentDivStyle.isNoFit && !commentsShowing ? 'black' : '',
            padding: 0,
            '&:last-child': { pb: 0 },
          }}
          >
            {!commentsShowing ? (
              <SwipableImage
                zoomInfo={zoomInfo}
                onSlidePin={onSlidePin}
                onSetImageMetaData={(m) => this.setState({ imgMetaData: m })}
              />
            )
              : (
                <Comments
                  stylingProps={commentsShowing}
                  pinInformation={pinInformation}
                  handleNewComment={handleNewComment}
                  authenticated={authenticated}
                  toggleComments={this.toggleComments}
                  closePin={(e: React.SyntheticEvent) => this.close(e, true)}
                  updateTags={updateTags}
                  displayLogin={displayLogin}
                />
              )}
          </CardContent>

        </Card>
      </>
    );
  }

}

export default PinZoom;
