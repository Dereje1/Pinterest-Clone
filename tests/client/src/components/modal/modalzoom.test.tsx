/**
 * @jest-environment jsdom
 */
import React from 'react';
import { EnzymePropSelector, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PinZoom, StyledBadge } from '../../../../../client/src/components/modal/modalzoom';
import { pinsStub } from '../../../stub';

jest.useFakeTimers();

describe('The pin zoom modal', () => {
  let props;
  const focus = jest.fn();
  beforeEach(() => {
    jest.spyOn(React, 'createRef').mockImplementation(() => ({
      current: {
        focus,
        clientHeight: 50,
        children: [{}, {
          clientHeight: 25,
        }],
      },
    }));
    props = {
      displayPinZoom: false,
      // [picobject, overlay button type, last scroll distance]
      zoomInfo: {
        pin: pinsStub[0],
        parentDivStyle: {
          top: 10,
          imgWidth: '622.5px',
          parentWidth: 622.5,
          pinnersSize: '2em',
          subTitleSize: '1.2em',
          titleSize: '2em',
          dateSize: '0.6em',
          width: '90%',
          isNoFit: false,
        },
      },
      reset: jest.fn(),
      pinImage: jest.fn(),
      deletePin: null,
      handleNewComment: jest.fn(),
      user: { authenticated: true },
      updateTags: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
    focus.mockClear();
  });

  test('will render', () => {
    const wrapper = shallow(<PinZoom {...props} />);
    expect(focus).toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will close the zoom window on blur', async () => {
    const wrapper = shallow(<PinZoom {...props} />);
    const zoomCard = wrapper.find({ className: 'zoom cshow' });
    wrapper.setState({ firstShow: false });
    zoomCard.props().onBlur();
    jest.advanceTimersByTime(500);
    await Promise.resolve();
    expect(props.reset).toHaveBeenCalledTimes(1);
  });

  test('will disable the scroll on mount and when user tries to scroll', () => {
    global.scrollTo = jest.fn();
    const mockedScroll = jest.mocked(global.scrollTo);
    const wrapper = shallow(<PinZoom {...props} />);
    mockedScroll.mockClear();
    const instance = wrapper.instance() as PinZoom;
    instance.disableScroll();
    expect(mockedScroll).toHaveBeenCalledWith(0, 10);
  });

  test('will toggle the comments window', () => {
    global.innerHeight = 1000;
    const updatedProps = {
      ...props,
      zoomInfo: {
        pin: pinsStub[0],
        parentDivStyle: {
          top: 10,
          imgWidth: '622.5px',
          parentWidth: 830,
          pinnersSize: '2em',
          subTitleSize: '1.2em',
          titleSize: '2em',
          dateSize: '0.6em',
          width: '90%',
          isNoFit: false,
        },
      },
    };
    const wrapper = shallow<PinZoom>(<PinZoom {...updatedProps} />);
    // toggle comment on
    let commentIcon: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    commentIcon = commentIcon.props().action.props.children[0].props.children.props;
    commentIcon.onClick();
    commentIcon.onMouseDown({ preventDefault: jest.fn() });
    expect(wrapper.state().commentsShowing).toEqual({ height: 950, width: 830 });
    expect(wrapper.state().cancelBlur).toEqual(true);
    // toggle comment off
    commentIcon.onClick();
    expect(wrapper.state().commentsShowing).toEqual(null);
    expect(wrapper.state().cancelBlur).toEqual(false);
  });

  test('will force close the comments window', () => {
    const wrapper = shallow(<PinZoom {...props} />);
    // toggle comment on
    let commentIcon: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    commentIcon = commentIcon.props().action.props.children[0].props.children.props;
    commentIcon.onClick();
    const commentsWindow: EnzymePropSelector = wrapper.find('Comments');
    // expect(wrapper.state().show).toBe(true);
    commentsWindow.props().closePin();
    // expect(wrapper.state().show).toBe(false);
  });

  test('will not close the comments window if blur is not cancelled', () => {
    const wrapper = shallow(<PinZoom {...props} />);
    // toggle comment on
    let commentIcon: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    commentIcon = commentIcon.props().action.props.children[0].props.children.props;
    // show comments div
    commentIcon.onClick();
    wrapper.setState({ cancelBlur: true });
    const instance = wrapper.instance() as PinZoom;
    instance.close({} as React.SyntheticEvent, false);
  });

  test('will set zero on badge content for pins icon if savedby is not defined', () => {
    const updatedProps = {
      ...props,
      zoomInfo: {
        ...props.zoomInfo,
        pin: {
          ...pinsStub[0],
          savedBy: undefined,
        },
      },
    };
    const wrapper = shallow(<PinZoom {...updatedProps} />);
    // toggle comment on
    let pinsIcon: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    pinsIcon = pinsIcon
      .props()
      .action
      .props
      .children[1]
      .props;
    expect(pinsIcon.badgeContent).toBe(0);
  });

  test('will add eventlisteners to listen for scrolling events on mount', () => {
    global.addEventListener = jest.fn();
    const mockedAddEventListener = jest.mocked(global.addEventListener);
    shallow(<PinZoom {...props} />);
    const events = mockedAddEventListener.mock.calls.map((c) => c[0]);
    expect(global.addEventListener).toHaveBeenCalledTimes(1);
    expect(events).toEqual(['scroll']);
  });

  test('will remove eventlisteners and reinstate scroll when modal unloads', () => {
    global.removeEventListener = jest.fn();
    const mockedRemoveEventListener = jest.mocked(global.removeEventListener);
    const wrapper = shallow(<PinZoom {...props} />);
    const instance = wrapper.instance() as PinZoom;
    instance.componentWillUnmount();
    const events = mockedRemoveEventListener.mock.calls.map((c) => c[0]);
    expect(global.removeEventListener).toHaveBeenCalledTimes(1);
    expect(events).toEqual(['scroll']);
  });

  test('will include a link to the profile of the owner in the subheader and close modal on click', async () => {
    const wrapper = shallow(<PinZoom {...props} />);
    const cardHeader: EnzymePropSelector = wrapper.find('ForwardRef(CardHeader)');
    const link = cardHeader.props().subheader.props.children[0].props;
    link.onMouseDown({ preventDefault: jest.fn() });
    link.onClick();
    jest.advanceTimersByTime(500);
    await Promise.resolve();
    expect(link.to).toBe('/profile/1');
    expect(props.reset).toHaveBeenCalledTimes(1);
  });
});

describe('The styled badge', () => {
  test('Will render for pins', () => {
    const badge = shallow(<StyledBadge name="pin" />);
    expect(toJson(badge)).toMatchSnapshot();
  });

  test('Will render for comments', () => {
    const badge = shallow(<StyledBadge name="comments" />);
    expect(toJson(badge)).toMatchSnapshot();
  });
});
