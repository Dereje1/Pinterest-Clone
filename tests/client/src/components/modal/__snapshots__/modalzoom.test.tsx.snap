// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The pin zoom modal will render 1`] = `
<Fragment>
  <div
    className="modal-overlay"
  />
  <ForwardRef(Card)
    className="zoom cshow"
    onBlur={[Function]}
    sx={
      {
        "top": 10,
        "width": 622.5,
      }
    }
    tabIndex={0}
  >
    <ForwardRef(CardHeader)
      action={
        <React.Fragment>
          <Styled(ForwardRef(Badge))
            badgeContent={1}
            color="primary"
            name="comments"
            showZero={true}
          >
            <ForwardRef(IconButton)
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <Memo
                style={
                  {
                    "fontSize": "1.7em",
                  }
                }
              />
            </ForwardRef(IconButton)>
          </Styled(ForwardRef(Badge))>
          <Styled(ForwardRef(Badge))
            badgeContent={2}
            color="secondary"
            name="pin"
            showZero={true}
          >
            <GetAction
              deletePin={null}
              element={
                {
                  "_id": "1",
                  "comments": [
                    {
                      "_id": "Comment_id_1",
                      "comment": "Test comment",
                      "createdAt": "2021-07-28T16:30:48.572Z",
                      "displayName": "tester-displayName",
                      "userId": "Comment_User_Id_1",
                    },
                  ],
                  "createdAt": "2020-12-25T00:00:00.000+00:00",
                  "hasSaved": false,
                  "imgDescription": "imgDescription id-1",
                  "imgLink": "imgLink id-1",
                  "owner": {
                    "name": "owner id-1",
                    "service": "google",
                    "userId": "1",
                  },
                  "owns": false,
                  "savedBy": [
                    {
                      "name": "savedBy - id-1",
                      "service": "google",
                      "userId": "1",
                    },
                    {
                      "name": "savedBy - id-1",
                      "service": "github",
                      "userId": "1",
                    },
                  ],
                  "tags": [
                    {
                      "_id": "tag_id_1",
                      "tag": "tag 1",
                    },
                    {
                      "_id": "tag_id_2",
                      "tag": "tag 2",
                    },
                  ],
                }
              }
              pinImage={[MockFunction]}
              reset={[Function]}
            />
          </Styled(ForwardRef(Badge))>
        </React.Fragment>
      }
      subheader={
        <React.Fragment>
          <ProfileLink
            authenticated={true}
            closePin={[Function]}
            displayLogin={[MockFunction]}
            title="owner id-1"
            userId="1"
          />
          <br />
          <span
            style={
              {
                "marginLeft": 0,
              }
            }
          >
            25 Dec 2020
          </span>
        </React.Fragment>
      }
      subheaderTypographyProps={
        {
          "fontWeight": "bold",
        }
      }
      title="imgDescription ..."
      titleTypographyProps={
        {
          "fontWeight": "bold",
        }
      }
    />
    <ForwardRef(CardContent)
      sx={
        {
          "&:last-child": {
            "pb": 0,
          },
          "background": "",
          "padding": 0,
        }
      }
    >
      <ForwardRef(CardMedia)
        component="img"
        id="pin-zoom"
        image="imgLink id-1"
        sx={
          {
            "marginLeft": "auto",
            "marginRight": "auto",
            "width": "622.5px",
          }
        }
      />
    </ForwardRef(CardContent)>
  </ForwardRef(Card)>
</Fragment>
`;

exports[`The styled badge Will render for comments 1`] = `
<ForwardRef(Badge)
  className="ForwardRef(Badge)-root-1 ForwardRef(Badge)-root-3"
  name="comments"
/>
`;

exports[`The styled badge Will render for pins 1`] = `
<ForwardRef(Badge)
  className="ForwardRef(Badge)-root-1 ForwardRef(Badge)-root-2"
  name="pin"
/>
`;
